### PR TITLE
feat: add MsgExecCompat and pick other minor changes

### PR DESCRIPTION
**Pull request by vinhphuctadang**
# Context
https://www.notion.so/injective/EIP712-compatible-MsgExec-22c134266b434f0caa8ebd407053ac57

# Changes

Create MsgExecCompat message that copies behavior from MsgExec, the only difference is to unmarshal JSON each message in msg[]

Also, we need to add `GlobalCdc` in order to parse proto JSON

# Test

Using client script to execute the message:
- Invalid message format:

```
import {
    getEthereumAddress,
    PrivateKey,
} from "@injectivelabs/sdk-ts";
import axios, { Axios } from "axios";
import https from "https";

const baseURL = "https://devnet.api.injective.dev";
const privateKey = PrivateKey.fromHex(
    "5d386fbdbf11f1141010f81a46b40f94887367562bd33b452bbaa6ce1cd1381e",
);
const injectiveAddress = privateKey.toAddress().toBech32();
const granterAddress = 'inj17gkuet8f6pssxd8nycm3qr9d9y699rupv6397z'
const ethereumAddress = getEthereumAddress(injectiveAddress);

export const prepareSignBroadcast = async (): Promise<unknown> => {
    const msgSend = {
        '@type': '/cosmos.bank.v1beta1.MsgSend',
        from_address: granterAddress,
        to_address: "inj1995xnrrtnmtdgjmx0g937vf28dwefhkhy6gy5e",
        amount: [
            {
                denom:"inj", amount:"10000000000000000"
            },
        ]
    }

    const msgAuthz = {
        grantee: 'inj14au322k9munkmx5wrchz9q30juf5wjgz2cfqku',
        msgs: [
            JSON.stringify(msgSend).substring(0, 15),
        ],
        '@type': '/cosmos.authz.v1beta1.MsgExecCompat'
    }

    console.log('msg to send:', msgAuthz)
    let client = new Axios({
        baseURL: baseURL,
        transformRequest: axios.defaults.transformRequest,
        transformResponse: axios.defaults.transformResponse,
        httpsAgent: new https.Agent({  
            rejectUnauthorized: false
        })
    })

    let msgJson = JSON.stringify(msgAuthz)
    let base64Msg = Buffer.from(msgJson).toString('base64')
    let prepareTxPayload = {
        "chainID": 5,
        "fee": {
        "delegateFee": false,
        "gas": 200000,
        "price": [
            {
                "amount": "10000000000000000000",
                "denom": "inj"
            }
        ]
        },
        "memo": "",
        "msgs": [
            base64Msg,
        ],
        "sequence": 0,
        "signerAddress": ethereumAddress,
        "timeoutHeight": 0
    }

    const prepareTxResponse = await client.post('/api/exchange/gateway/v1/prepareTx', prepareTxPayload);
    const prepareTxResponseBody = prepareTxResponse.data
    console.log('prepareTx response:', prepareTxResponseBody)

    const signature = await privateKey.signTypedData(
        JSON.parse(prepareTxResponseBody.data),
    );
    
    const parsedTypeData = JSON.parse(prepareTxResponseBody.data)
    const tx = parsedTypeData.message
    const msgs = [
    base64Msg,
    ]
    
    const broadcastPayload = {
    "chainID": 5,
    "feePayer": prepareTxResponseBody.feePayer,
    "feePayerSig": prepareTxResponseBody.feePayerSig,
    "msgs": msgs,
    "pubKey": {
        "key": '0x' + privateKey.toPublicKey().toHex(),
        "type": "/injective.crypto.v1beta1.ethsecp256k1.PubKey"
    },
    "signature": '0x' + Buffer.from(signature).toString('hex'),
    "tx": Buffer.from(JSON.stringify(tx)).toString('base64'),
    "mode": 'block',
    }
    const broadcastResponse = await client.post('/api/exchange/gateway/v1/broadcastTx', broadcastPayload)
    console.log('broadcastTx response:', broadcastResponse.data)
    return '' // broadcastResponse.data;
};

(async () => {
    await prepareSignBroadcast();
})();

```

Error happens:

image

Using valid message format:

```
import {
    getEthereumAddress,
    PrivateKey,
} from "@injectivelabs/sdk-ts";
import axios, { Axios } from "axios";
import https from "https";

// const baseURL = "https://localhost:4448";
const baseURL = "https://devnet.api.injective.dev";
const privateKey = PrivateKey.fromHex(
    "5d386fbdbf11f1141010f81a46b40f94887367562bd33b452bbaa6ce1cd1381e",
);
const injectiveAddress = privateKey.toAddress().toBech32();
const granterAddress = 'inj17gkuet8f6pssxd8nycm3qr9d9y699rupv6397z'
const ethereumAddress = getEthereumAddress(injectiveAddress);

export const prepareSignBroadcast = async (): Promise<unknown> => {
    const msgSend = {
        '@type': '/cosmos.bank.v1beta1.MsgSend',
        from_address: granterAddress,
        to_address: "inj1995xnrrtnmtdgjmx0g937vf28dwefhkhy6gy5e",
        amount: [
            {
                denom:"inj", amount:"10000000000000000"
            },
        ]
    }

    const msgCreateDenom = {
        '@type': '/injective.tokenfactory.v1beta1.MsgCreateDenom',
        'sender': granterAddress,
        'subdenom': 'exampledenom'
    }

    const msgAuthz = {
        grantee: 'inj14au322k9munkmx5wrchz9q30juf5wjgz2cfqku',
        msgs: [
            JSON.stringify(msgSend),
            JSON.stringify(msgCreateDenom),
        ],
        '@type': '/cosmos.authz.v1beta1.MsgExecCompat'
    }

    console.log('msg to send:', msgAuthz)
    let client = new Axios({
        baseURL: baseURL,
        transformRequest: axios.defaults.transformRequest,
        transformResponse: axios.defaults.transformResponse,
        httpsAgent: new https.Agent({  
            rejectUnauthorized: false
        })
    })

    let msgJson = JSON.stringify(msgAuthz)
    let base64Msg = Buffer.from(msgJson).toString('base64')
    let prepareTxPayload = {
        "chainID": 5,
        "fee": {
        "delegateFee": false,
        "gas": 200000,
        "price": [
            {
                "amount": "10000000000000000000",
                "denom": "inj"
            }
        ]
        },
        "memo": "",
        "msgs": [
            base64Msg,
        ],
        "sequence": 0,
        "signerAddress": ethereumAddress,
        "timeoutHeight": 0
    }

    const prepareTxResponse = await client.post('/api/exchange/gateway/v1/prepareTx', prepareTxPayload);
    const prepareTxResponseBody = prepareTxResponse.data
    console.log('prepareTx response:', prepareTxResponseBody)

    const signature = await privateKey.signTypedData(
        JSON.parse(prepareTxResponseBody.data),
    );
    
    const parsedTypeData = JSON.parse(prepareTxResponseBody.data)
    const tx = parsedTypeData.message
    const msgs = [
    base64Msg,
    ]
    
    const broadcastPayload = {
    "chainID": 5,
    "feePayer": prepareTxResponseBody.feePayer,
    "feePayerSig": prepareTxResponseBody.feePayerSig,
    "msgs": msgs,
    "pubKey": {
        "key": '0x' + privateKey.toPublicKey().toHex(),
        "type": "/injective.crypto.v1beta1.ethsecp256k1.PubKey"
    },
    "signature": '0x' + Buffer.from(signature).toString('hex'),
    "tx": Buffer.from(JSON.stringify(tx)).toString('base64'),
    "mode": 'block',
    }
    const broadcastResponse = await client.post('/api/exchange/gateway/v1/broadcastTx', broadcastPayload)
    console.log('broadcastTx response:', broadcastResponse.data)
    return '' // broadcastResponse.data;
};

(async () => {
    await prepareSignBroadcast();
})();
```

According to events, messages are executed correctly:

image

image

image

### CODE REVIEW

File: x/authz/msgs.go [hunk 1]
```diff
@@ -1,9 +1,11 @@
 package authz
 
 import (
+	"fmt"
 	"time"
 
 	"github.com/cosmos/cosmos-sdk/x/auth/migrations/legacytx"
+	"github.com/cosmos/cosmos-sdk/x/authz/codec"
```
Comment (dbrajovic) [inline x/authz/msgs.go]: Why the double import?

File: baseapp/baseapp.go [hunk 2]
```diff
@@ -144,6 +144,9 @@ type BaseApp struct { //nolint: maligned
 	abciListeners []ABCIListener
 
 	chainID string
+
+	// StreamEvents
+	StreamEvents chan StreamEvents
```
Comment (dbrajovic) [inline baseapp/baseapp.go:149]: Where is this channel drained (read from)?

File: x/authz/codec/cdc.go [hunk 3]
```diff
@@ -2,17 +2,20 @@ package codec
...
 	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/codec/types"
 	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
 var (
 	Amino     = codec.NewLegacyAmino()
 	ModuleCdc = codec.NewAminoCodec(Amino)
+	GlobalCdc *codec.ProtoCodec
```
Comment (dbrajovic) [inline x/authz/codec/cdc.go]: You can inline this as:

`ProtoCdc = codec.NewProtoCodec(types.NewInterfaceRegistry())`

**Review by albertchon — APPROVED** [review]

File: baseapp/baseapp.go (hunk 2 above)
Comment (vinhphuctadang) [reply baseapp/baseapp.go:149]: It's read from injective-core > event stream

**Review by dbrajovic — APPROVED** [review]

**Pull request merged**